Fix problems with some hack files (for example Trig.hack)

scalvin1 opened the pull request:
Fixes 2 problems:

- Handling of Windows-type CR_LF line endings in some .hack files (such as Trig.hack)
- Keys not working in some .hack files (lower to upper case keycodes fixes that)

scalvin1 committed:
Some debugging output, not needed

scalvin1 committed:
Fix problem with some .hack files:
ROM loader got confused by stupid windows CR LF line endings. Extended the buffer a little now.

scalvin1 committed:
translate lower key input values to upper key (CPU Emulator from Nand2Tetris does this)
Fixes some .hack files not running correctly

scalvin1 committed:
Make it run as fast as possible (at least on my old notebook)

kurtjd commented:
Thanks for this! You saved me a lot of hours of debugging haha, I didn't even to begin to suspect Windows' newline character to be the issue.

Though I do have a couple nitpicky things/questions:
* Do you mind removing the commented-out line of codes that I assume you were using for debugging?
* I see you increased the buffer size for lines from WORD_SIZE + 2 to WORD_SIZE + 5 to accommodate the extra \\r character. However, testing this with just + 3 seemed to work fine, I think because \\r\\n\\0 is actually 3 characters.
* You commented out the speed capping condition. <del>Did you do this intentionally? If not, please keep that line in unless you think speed shouldn't be capped (at least, in my experience, Pong was unplayable without an artificial cap).</del> I looked at the commit history and saw you did comment out that line intentionally, but I'd still prefer you keep it in because certain ROMs such as Pong are unplayable at uncapped speeds and users can instead adjust the CPU_FREQ definition if they want to speed it up. I can also look into adding a command-line argument to set/uncap CPU speed.

**Edit**: I might be able to merge the PR with only cherry-picked commits since you helpfully made different commits for different modifications. I'll see if I can do that to save you some time.

Thanks!

scalvin1 committed:
Limit changes to actual bug fixes. This fixes:
- a-z Keys not recognized the same as in CPU Emulator
- CR-LF Hack ROM files not loading correctly (were adding @0 instructions everywhere)

scalvin1 commented:
I am completely new to these pull-requests and I hoped this would be easier picking the relevant diff stubs.
But no worries, I changed it to reflect only the two little fixes now. Otherwise- commits are free, just change it. Or close the PR and apply the fix yourself. Not bothered at all, I just thought I should let you know that I am enjoying your program and fixed some small issues. :-)
Best regards,
Jan

kurtjd commented:
Thank you! Sorry about that, I'm pretty sure there is a way of cherry-picking commits on my end but I'm not terribly familiar with git beyond committing, branching, merging, and pushing haha.

But again I appreciate it, and if you ever run into any other issues or ideas for improvements, feel free to let me know or submit future pull requests.
